File: src/main/java/kr/co/lunatalk/global/config/security/CustomAccessDeniedHandler.java
package kr.co.lunatalk.global.config.security;

import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import kr.co.lunatalk.global.common.response.ErrorResponse;
import kr.co.lunatalk.global.common.response.GlobalResponse;
import kr.co.lunatalk.global.exception.ErrorCode;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.web.access.AccessDeniedHandler;
import org.springframework.stereotype.Component;

import java.io.IOException;

@Component
@RequiredArgsConstructor
public class CustomAccessDeniedHandler implements AccessDeniedHandler {

	private final ObjectMapper mapper;

	@Override
	public void handle(HttpServletRequest request, HttpServletResponse response, AccessDeniedException accessDeniedException) throws IOException, ServletException {
		response.setContentType("application/json;charset=UTF-8");
		response.setStatus(HttpStatus.FORBIDDEN.value());
//		ObjectMapper mapper = new ObjectMapper().registerModule(new JavaTimeModule()).disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
		GlobalResponse errorResponse = GlobalResponse.fail(ErrorCode.AUTH_FAILED.getHttpStatus().value(), new ErrorResponse(ErrorCode.AUTH_FAILED.name(), ErrorCode.AUTH_FAILED.getMessage()));
		response.getWriter().write(mapper.writeValueAsString(errorResponse));
	}
}

File: lunatalk-api/src/main/kotlin/kr/co/lunatalk/global/config/security/CustomAccessDeniedHandler.kt
package kr.co.lunatalk.global.config.security

import tools.jackson.databind.ObjectMapper
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import kr.co.lunatalk.global.common.response.ErrorResponse
import kr.co.lunatalk.global.common.response.GlobalResponse
import kr.co.lunatalk.global.exception.ErrorCode
import kr.co.lunatalk.global.util.CookieUtil
import org.springframework.context.annotation.Lazy
import org.springframework.http.HttpStatus
import org.springframework.security.access.AccessDeniedException
import org.springframework.security.web.access.AccessDeniedHandler
import org.springframework.stereotype.Component

@Component
class CustomAccessDeniedHandler(
    @Lazy private val mapper: ObjectMapper,
    @Lazy private val cookieUtil: CookieUtil
) : AccessDeniedHandler {

    override fun handle(
        request: HttpServletRequest,
        response: HttpServletResponse,
        accessDeniedException: AccessDeniedException
    ) {
        cookieUtil.deleteAccessTokenCookie(response)

        response.contentType = "application/json;charset=UTF-8"
        response.status = HttpStatus.FORBIDDEN.value()
        val errorResponse = GlobalResponse.fail(
            ErrorCode.AUTH_FAILED.httpStatus.value(),
            ErrorResponse(ErrorCode.AUTH_FAILED.name, ErrorCode.AUTH_FAILED.message)
        )
        response.writer.write(mapper.writeValueAsString(errorResponse))
    }
}
